### ISSUE TITLE
Purpose of padding

### ISSUE THREAD

**Issue by AlessandroPolidori** [nelson1425/EfficientAD#12]
https://github.com/nelson1425/EfficientAD/blob/fcab5146f84ae17597044ad5ddf1656ccf805401/efficientad.py#LL279C9-L279C75
I'm wondering why this padding is necessary. Can you help me?

**Comment by nelson1425** [comment]
Hi, without this padding the segmentation metric will decrease very much because the segmentation pixels are not aligned anymore after the next line (map_combined = torch.nn.functional.interpolate(...)).

It is similar to applying 11x11 edge filter to 100x100 image. The result will be 90x90 image with detected edges. Resizing it to 100x100 would distort the detected edges. Instead you have to pad the 90x90 image with 5 pixels on each size.

In the case of EfficientAD, the patch size is 33, so 16 pixels padding on each border. The teacher has two layers with stride 2, so we need to divide the 16 pixels padding by 2 and again by 2, which gives us 4. So 4 pixels padding on each border for the output feature maps.

**Comment by jpcbertoldo** [comment]
> disclaimer: i'm using the implementation in [anomalib](https://github.com/openvinotoolkit/anomalib/blob/c97762c8781aacfc31d100ba862aa730a3262f7a/src/anomalib/models/efficient_ad/torch_model.py), which apparently was based on yours
>
> there is an option `pad_maps` to use this padding you guys are talking about

@nelson1425 i don't think this is correct, because if you add a 0 padding in the end then upscale it with an interpolation, you would be just upscaling the zero borders as well, while just upscaling would actually make the corners at the final resolution align (obs: 56 is the output resolution without any padding with the small PDN, the blue zone is filled with 0s): 

![image](https://github.com/nelson1425/EfficientAD/assets/24547377/eafa3ab8-fb8c-4cd1-aa20-8212e3b8e739)
 
i checked this with the implementation in anomalib and it's actually the case (a random score map from an untrained model): 

![image](https://github.com/nelson1425/EfficientAD/assets/24547377/d36d79f7-13fe-41e6-90d1-99bab284c4b6)

---

as a result,  that null score zone is independent of the input, so i'd expect results to be worse (?) , so 

> ... without this padding the segmentation metric will decrease very much ...

sounds rather surprinsing; did you test this factor isolated?
 
besides, the paper says 

![image](https://github.com/nelson1425/EfficientAD/assets/24547377/44322a63-7efb-4079-84c4-405bb808c8cf)

(appendix E, page 21)

so the reported results are without padding, and they don't say anything about this zero padding in the end, and it supposedly shouldn't affect much the results

**Comment by limaodaxia** [comment]
> Hi, without this padding the segmentation metric will decrease very much because the segmentation pixels are not aligned anymore after the next line (map_combined = torch.nn.functional.interpolate(...)).
> 
> It is similar to applying 11x11 edge filter to 100x100 image. The result will be 90x90 image with detected edges. Resizing it to 100x100 would distort the detected edges. Instead you have to pad the 90x90 image with 5 pixels on each size.
> 
> In the case of EfficientAD, the patch size is 33, so 16 pixels padding on each border. The teacher has two layers with stride 2, so we need to divide the 16 pixels padding by 2 and again by 2, which gives us 4. So 4 pixels padding on each border for the output feature maps.

![image](https://github.com/nelson1425/EfficientAD/assets/68214694/461ab63d-489c-4ee1-a444-97788bc03455)

how caculate these values? how can I know what size should I pad?
For example, when it comes to these origin sizes, what size should I pad if I want to resize 56×56 to these origin sizes?
